src/sentry/workflow_engine/processors/delayed_workflow.py

```diff
@@ -410,13 +412,18 @@ def fire_actions_for_groups(
                 action_filters.add(dcg)
 
         # process action filters
-        filtered_actions: list[Action] = list(
-            filter_recently_fired_workflow_actions(action_filters, group)
-        )
+        filtered_actions = filter_recently_fired_workflow_actions(action_filters, group)
+        # update WorkflowFireHistory for the filtered actions (created before enqueuing)
+        update_workflow_fire_histories(filtered_actions, event_data)
 
         # process workflow_triggers
         workflows = set(Workflow.objects.filter(when_condition_group_id__in=workflow_triggers))
-        filtered_actions.extend(list(evaluate_workflows_action_filters(workflows, job)))
+
+        # create WorkflowFireHistory (updated in evaluate_workflows_action_filters)
+        create_workflow_fire_histories(workflows, event_data)
```

Comment:
this feels like a side effect of `fire_actions_for_groups` is there another place we could do this? The way i'd structure the code is to evaluate the actions, return the list of actions, then have another method trigger them (no queries or anything in that firing method, just taking a list of actions and calling anything that needs to be fired)

Reply:
i agree that ideally this would not be here and preferably would be done beforehand...

logically it belongs right after this in `get_groups_to_fire` https://github.com/getsentry/sentry/blob/14f7a70fdf9cf2fa666da33f9faea83d959d2d05/src/sentry/workflow_engine/processors/delayed_workflow.py#L289

i'm not certain if it's possible to nicely slot it in `get_groups_to_fire` because:
1) `WorkflowFireHistory` requires the `event_id` from the `GroupEvent`
2) we grab the `GroupEvents` in bulk after exiting from `get_groups_to_fire` (aka after we figure out which groups we need to fire workflows for)

given this, would it be okay to leave it where it is? i've refactored away other standalone usages save for this one

Reply:
Oof yeah this is tricky, I looked at it for a bit trying to figure out another way and I can't think of anything super clean. You _could_ store `filtered_actions` in 2 separate variables (since you tack on the results of `evaluate_workflows_action_filters` and wouldn't want that) and return both along with the corresponding event ids (or just the event ids ins separate vars?) so you can then call `create_workflow_fire_histories` in `process_delayed_workflows`, but idk if that's really much better than you have it now.